File: lib/schema/state.ts
export type State = Partial<{
	Calculator: Calculator,
	Members: null | Member[],
}>;

export type Member = {
    Id: number,
    X: number,
    Y: number,
};

export type Calculator = {
    // row 0
    Equation: string,
	Display: string,
    // row 1
    Clear: boolean,
    Bracket: boolean,
    Percentage: boolean,
    Divide: boolean,
    // row 2
    Seven: boolean,
    Eight:  boolean,
    Nine: boolean,
    Times: boolean,
    // row 3
    Four: boolean,
    Five: boolean,
    Six: boolean,
    Minus: boolean,
    // row 4
    One: boolean,
    Two: boolean,
    Three: boolean,
    Plus: boolean,
    // row 5
    Negate: boolean,
    Zero: boolean,
    Decimal: boolean,
    Equals: boolean,
};

File: lib/schema/state.go
package schema

type PersonalizedState struct {
	Id    int
	State State
}

type State struct {
	Calculator Calculator
	Members    []Member
}

type StateState struct {
	Calculator Calculator
	Members    []MemberState
}

type Member struct {
	Id int
	X  int
	Y  int
}

type MemberState struct {
	Member Member
	// row 1
	Clear      bool
	Bracket    bool
	Percentage bool
	Divide     bool
	// row 2
	Seven bool
	Eight bool
	Nine  bool
	Times bool
	// row 3
	Four  bool
	Five  bool
	Six   bool
	Minus bool
	// row 4
	One   bool
	Two   bool
	Three bool
	Plus  bool
	// row 5
	Negate  bool
	Zero    bool
	Decimal bool
	Equals  bool
}

type Calculator struct {
	// row 0
	Equation string
	Display  string
	// row 1
	Clear      bool
	Bracket    bool
	Percentage bool
	Divide     bool
	// row 2
	Seven bool
	Eight bool
	Nine  bool
	Times bool
	// row 3
	Four  bool
	Five  bool
	Six   bool
	Minus bool
	// row 4
	One   bool
	Two   bool
	Three bool
	Plus  bool
	// row 5
	Negate  bool
	Zero    bool
	Decimal bool
	Equals  bool
}

func (s *StateState) State() State {
	members := make([]Member, len(s.Members))
	for idx, member := range s.Members {
		members[idx] = member.Member
	}
	return State{
		Calculator: s.Calculator,
		Members:    members,
	}
}

func (s *State) WithoutMember(id int) State {
	members := make([]Member, 0, len(s.Members)-1)
	for _, member := range s.Members {
		if member.Id != id {
			members = append(members, member)
		}
	}
	return State{
		Calculator: s.Calculator,
		Members:    members,
	}
}
